guides/c.md

```diff
@@ -1,10 +1,98 @@
-The C wrappers should look like the C++ code as much as possible. But, there are going to be differences between the wrappers and the C++ code for many reasons.
+Introduction
+============
 
-- The projects we are wrapping usually have a style guide. Follow that, or make a best guess based on the code already written.
-- Usually the projects use clang-format. It's highly recommended to also use clang-format, and run it whenever the file is saved.
-- In the case of overrides, check if the override is a convenience override or not. If it is not, then ... (TODO: what guideline should be followed?)
+The C wrappers should look like the C++ code as much as possible. But, there are going to be differences between the wrappers and the C++ code for many reasons. When in doubt, refer to to the wrapped library's style guides, or make a best guess based on code already written in either the wrapped library, or other wrappers.
 
-Here are some examples of how to style the C code.
+Formatting
+==========
+
+Use autoformatting where possible. Generally the wrapped libraries use clang-format, so it's best to use that. Also, it is suggested to do formatting on save.
+
+Implementation
+==============
+
+- If there are function overrides, then prefer to wrap the core implementation. If there's multiple implementations, then try to pick a suffix that is clear and understandable.
+
+Error Handling
+==============
+
+TODO
+
+SUGGESTION: Implement something like Rust's `Result<T, E>`, unless the wrapped library already implements error handling that we can inherit from in C.
+
+```c
+enum StatusCode {
+    OK;
+    ERROR;
+};
+typedef struct {
+    StatusCode status;
+    union value {
+        MyType value;
+        Error err;
+    }
+} MyTypeStatus;
+```
+
+Exceptions
+==========
+
+C does not support exceptions, and it's generally considered bad to allow exceptions to pass across the C++ to C boundary (TODO: Why?). So, if wrapped code can cause an exception, then it should be wrapped in code like below.
```

Comment:
Having an exception fire across FFI in Rust is undefined behaviour. I'm not sure what the situation is with C but I would assume the same. At the very least, if the C++ library in question is relying on exceptions unwinding the stack correctly in order to clean up resources (open file handles/sockets etc), those will be left in a bad state

Reply:
Upon further reading it looks like the actual answer is more complicated and somewhat platform-specific: https://stackoverflow.com/questions/4448677/can-a-c-program-handle-c-exceptions 

But yeah, I think the advice should strongly be to handle any exceptions in the wrapping function (which requires inspecting the C++ code carefully to see what might bubble up).

Reply:
I'll probably call it undefined behaviour then.

Reply:
Slightly unrelated, but it looks like we could do something like this in our c++ side of the wrapper to prevent exceptions from being thrown:

```cpp
void no_throw();
void may_throw();

static_assert(noexcept(no_throw()));
static_assert(noexcept(may_throw()));
```

Reply:
> Slightly unrelated, but it looks like we could do something like this in our c++ side of the wrapper to prevent exceptions from being thrown:
> 
> ```c++
> void no_throw();
> void may_throw();
> 
> static_assert(noexcept(no_throw()));
> static_assert(noexcept(may_throw()));
> ```

 Are you saying that all wrapped methods/functions will have to be noexcept? I'm not sure that's going to be possible.

Reply:
I could be wrong, but that could work on the cpp side of the wrapper. If this doesn't work, or doesn't produce the expected results, then this is safe to ignore.